hack/generator/pkg/jsonast/jsonast.go

```diff
@@ -604,98 +608,21 @@ func oneOfHandler(ctx context.Context, scanner *SchemaScanner, schema Schema) (a
-			}
-			propertyName := scanner.idFactory.CreatePropertyName(concreteType.Name(), astmodel.Exported)
-
-			// JSON name is unimportant here because we will implement the JSON marshaller anyway,
-			// but we still need it for controller-gen
-			jsonName := scanner.idFactory.CreateIdentifier(concreteType.Name(), astmodel.NotExported)
-			property := astmodel.NewPropertyDefinition(
-				propertyName, jsonName, concreteType).MakeOptional().WithDescription(propertyDescription)
-			properties = append(properties, property)
-		case *astmodel.EnumType:
-			// TODO: This name sucks but what alternative do we have?
-			name := fmt.Sprintf("enum%v", i)
-			propertyName := scanner.idFactory.CreatePropertyName(name, astmodel.Exported)
-
-			// JSON name is unimportant here because we will implement the JSON marshaller anyway,
-			// but we still need it for controller-gen
-			jsonName := scanner.idFactory.CreateIdentifier(name, astmodel.NotExported)
-			property := astmodel.NewPropertyDefinition(
-				propertyName, jsonName, concreteType).MakeOptional().WithDescription(propertyDescription)
-			properties = append(properties, property)
-		case *astmodel.ObjectType:
-			// TODO: This name sucks but what alternative do we have?
-			name := fmt.Sprintf("object%v", i)
-			propertyName := scanner.idFactory.CreatePropertyName(name, astmodel.Exported)
-
-			// JSON name is unimportant here because we will implement the JSON marshaller anyway,
-			// but we still need it for controller-gen
-			jsonName := scanner.idFactory.CreateIdentifier(name, astmodel.NotExported)
-			property := astmodel.NewPropertyDefinition(
-				propertyName, jsonName, concreteType).MakeOptional().WithDescription(propertyDescription)
-			properties = append(properties, property)
-		case *astmodel.PrimitiveType:
-			var primitiveTypeName string
-			if concreteType == astmodel.AnyType {
-				primitiveTypeName = "anything"
-			} else {
-				primitiveTypeName = concreteType.Name()
-			}
-
-			// TODO: This name sucks but what alternative do we have?
-			name := fmt.Sprintf("%v%v", primitiveTypeName, i)
-			propertyName := scanner.idFactory.CreatePropertyName(name, astmodel.Exported)
-
-			// JSON name is unimportant here because we will implement the JSON marshaller anyway,
-			// but we still need it for controller-gen
-			jsonName := scanner.idFactory.CreateIdentifier(name, astmodel.NotExported)
-			property := astmodel.NewPropertyDefinition(
-				propertyName, jsonName, concreteType).MakeOptional().WithDescription(propertyDescription)
-			properties = append(properties, property)
-		default:
-			return nil, errors.Errorf("unexpected oneOf member, type: %T", t)
-		}
-	}
-
-	objectType := astmodel.NewObjectType().WithProperties(properties...)
-	objectType = objectType.WithFunction(
-		astmodel.JSONMarshalFunctionName,
-		astmodel.NewOneOfJSONMarshalFunction(objectType, scanner.idFactory))
-
-	return objectType, nil
+	return astmodel.MakeOneOfType(types), nil
```

Comment:
Also yay! (I love it when you get a place where the point of the PR is super clear. It's like when they say the name of the movie in the movie.)